### ISSUE TITLE
mastodon-android crashes after authentication

### ISSUE THREAD

**Issue by mishnz** [mastodon/mastodon-android#483]
Once authenticated mastodon-android stays logged in, but proceeds no further and crashes.
Upon launching mastodon-android again it instantly crashes after attempting to retrieve and decode data from the server.
The exception is:
```
12-31 09:45:47.517 22522 22541 E AndroidRuntime: FATAL EXCEPTION: MastodonAPIController
12-31 09:45:47.517 22522 22541 E AndroidRuntime: Process: org.joinmastodon.android, PID: 22522
12-31 09:45:47.517 22522 22541 E AndroidRuntime: java.lang.IllegalArgumentException: bad base-64
12-31 09:45:47.517 22522 22541 E AndroidRuntime: at android.util.Base64.decode(Base64.java:163)
12-31 09:45:47.517 22522 22541 E AndroidRuntime: at android.util.Base64.decode(Base64.java:138)
12-31 09:45:47.517 22522 22541 E AndroidRuntime: at android.util.Base64.decode(Base64.java:120)
12-31 09:45:47.517 22522 22541 E AndroidRuntime: at org.joinmastodon.android.api.PushSubscriptionManager$a.lambda$onSuccess$0(:1)
12-31 09:45:47.517 22522 22541 E AndroidRuntime: at org.joinmastodon.android.api.PushSubscriptionManager$a.a(Unknown Source:0)
12-31 09:45:47.517 22522 22541 E AndroidRuntime: at org.joinmastodon.android.api.k0.run(Unknown Source:4)
12-31 09:45:47.517 22522 22541 E AndroidRuntime: at android.os.Handler.handleCallback(Handler.java:942)
12-31 09:45:47.517 22522 22541 E AndroidRuntime: at android.os.Handler.dispatchMessage(Handler.java:99)
12-31 09:45:47.517 22522 22541 E AndroidRuntime: at android.os.Looper.loopOnce(Looper.java:226)
12-31 09:45:47.517 22522 22541 E AndroidRuntime: at android.os.Looper.loop(Looper.java:313)
12-31 09:45:47.517 22522 22541 E AndroidRuntime: at a0.j.run(:6)
```

Happy to run a debug version to isolate the exact base-64 data at fault if you can provide one.

**Comment by mishnz** [comment]
I've proxied the application and on every launch (including the first where it authenticates) it crashes.
The first launch (with initial authentication) makes the following HTTP requests (in sequence) before crash:
```
https://android.apis.google.com/c2dm/register3
https://reign.appmeasurements.com/ethic/onward
https://api.github.com/repos/mastodon/mastodon-android/releases/latest
https://api.joinmastodon.org/servers
https://my-mastodon-instance/api/v1/instance
https://my-mastodon-instance/api/v1/apps
https://my-mastodon-instance/oauth/token
https://my-mastodon-instance/api/v1/accounts/verify_credentials
https://my-mastodon-instance/api/v1/custom_emojis
https://my-mastodon-instance/api/v1/push/subscription
https://my-mastodon-instance/api/v1/filters
https://my-mastodon-instance/avatars/original/missing.png
https://my-mastodon-instance/api/v1/timelines/home?limit=40
https://my-mastodon-instance/headers/original/missing.png

```
On the next run (now that it's authenticated it only requests the following before crashing:
```
https://reign.appmeasurements.com/ethic/onward
https://my-mastodon-instance/api/v1/timelines/home?limit=40
https://my-mastodon-instance/api/v1/push/subscription
```
This means it has either cached something locally from the first run that is causing issues, or the "bad base-64" error is happening when it parses either "/api/v1/timelines/home?limit=40" or "/api/v1/push/subscription".

**Comment by mishnz** [comment]
Since the base64 code is all in /mastodon/src/main/java/org/joinmastodon/android/api/PushSubscriptionManager.java that really does point to something in the response of "/api/v1/push/subscription" triggering this crash.

**Comment by mishnz** [comment]
If this is the case, then I would assume that somehow the base64 data in the public "server_key" is causing the crash when it is being decoded, this reads as:
```
{
  "id": 229,
  "endpoint": "https://app.joinmastodon.org/relay-to/fcm/eRDk-trIY5o:APA91bGTw0Plicw3lNlA9Jt4O0i1bc74MGGi6v3YdC-6f-UbowNUkwI60JvoUlyKbdfmNfHKTISWjLOb4CWSn5DCRUkDqQobqb_VrZ2y4a5NliAZnxh9qaIsxtmRKcYsSTuC7W5yr2Mc/DUcl-RWqigdBfAONfeFcsg",
  "alerts": {
    "mention": true,
    "reblog": true,
    "follow": true,
    "favourite": true,
    "poll": true
  },
  "server_key": "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEPc4bZE6kXAz1qr/1CTnulqeqP8CGPFPTy9bsIwVL0qSPZjqw6afvOW55zcLdocleUppW6/CUOdC2oS8yZqpyVQ=="
}
```

**Comment by mishnz** [comment]
This would then mean the following line in PushSubscriptionManager.java could be causing the crash when reading in the server_key above:
`serverKey=deserializeRawPublicKey(Base64.decode(result.serverKey, Base64.URL_SAFE));`

**Comment by mishnz** [comment]
I wonder if this is down to the fact that the "server_key" contains forward slashes, but mastodon-android is using URL_SAFE which expects forward slashes to have been replaced by underscore characters.

**Comment by mishnz** [comment]
Confirmed that replacing the line above with:
`serverKey=deserializeRawPublicKey(Base64.decode(result.serverKey, Base64.NO_WRAP | Base64.NO_PADDING));
`and recompiling the apk resolves the issue.
I'll dig a little deeper and see if there are cases where Base64.URL_SAFE is actually required for the serverKey string.

**Comment by mishnz** [comment]
After testing and examining the Mastodon source code, Mastodon does **not** generate the serverKey using URL_SAFE.  This means that when the Mastodon Android client tries to pass a non URL_SAFE serverKey it will crash if it encounters a "/" or "+" as these are valid in the DEFAULT standard, but not in URL_SAFE.  Since Mastodon does not generate URL_SAFE serverKeys changing Base64.URL_SAFE to Base64.DEFAULT corrects the issue.   Tested and pull request generated in https://github.com/mastodon/mastodon-android/pull/486

**Comment by mishnz** [comment]
Leaving this open until the pull request is approved.

**Comment by mishnz** [comment]
> @[Gargron](https://github.com/Gargron) merged commit [525cc69](https://github.com/mastodon/mastodon-android/commit/525cc69c703b3b850eee23c1394e429b805fac9c) into mastodon:master [13 hours ago](https://github.com/mastodon/mastodon-android/pull/486#event-8133675864)

Thanks @[Gargron](https://github.com/Gargron), closing issue. :)
